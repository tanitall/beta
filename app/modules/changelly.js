"use strict";
module.exports = (function() {
  let URL = "https://api.changelly.com";
  let io = require("socket.io-client");
  let jayson = require("jayson");
  let crypto = require("crypto");
  let client = jayson.client.https(URL);

  function Changelly(apiKey, apiSecret) {
    this._id = function() {
      return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, function(
        c
      ) {
        let r = (Math.random() * 16) | 0,
          v = c == "x" ? r : (r & 0x3) | 0x8;
        return v.toString(16);
      });
    };

    this._sign = function(message) {
      return crypto
        .createHmac("sha512", apiSecret)
        .update(JSON.stringify(message))
        .digest("hex");
    };

    this._request = function(method, options, callback) {
      let id = this._id();
      let message = jayson.utils.request(method, options, id);
      client.options.headers = {
        "api-key": apiKey,
        sign: this._sign(message)
      };

      client.request(method, options, id, function(err, response) {
        callback(err, response);
      });
    };

    let self = this;

    this._socket = io.connect(URL, {
      reconnection: true,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
      reconnectionAttempts: "Infinity"
    });

    this._socket.on("connect", function() {
      let message = {
        Login: {}
      };

      self._socket.emit("subscribe", {
        apiKey,
        sign: self._sign(message),
        message
      });
    });
  }

  Changelly.prototype = {
    getCurrencies(callback) {
      return this._request("getCurrencies", {}, callback);
    },
    createTransaction(from, to, address, amount, extraId, callback) {
      let params = {
        from,
        to,
        address,
        amount,
        extraId
      };

      return this._request("createTransaction", params, callback);
    },
    getMinAmount(from, to, callback) {
      let params = {
        from,
        to
      };

      return this._request("getMinAmount", params, callback);
    },
    getExchangeAmount(from, to, amount, callback) {
      let params = {
        from,
        to,
        amount
      };

      return this._request("getExchangeAmount", params, callback);
    },
    getTransactions(limit, offset, currency, address, extraId, callback) {
      let params = {
        limit,
        offset,
        currency,
        address,
        extraId
      };

      return this._request("getTransactions", params, callback);
    },
    getStatus(id, callback) {
      let params = {
        id
      };

      return this._request("getStatus", params, callback);
    },
    on(channel, callback) {
      this._socket.on(channel, callback);
    }
  };

  return Changelly;
})();
