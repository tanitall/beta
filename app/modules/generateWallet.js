import {
  getAccountsFromWIFKey,
  generatePrivateKey,
  getWIFFromPrivateKey,
  encrypt_wif
} from "neon-js";

// Constants
export const NEW_WALLET_KEYS = "NEW_WALLET_KEYS";
export const NEW_WALLET = "NEW_WALLET";
export const SET_GENERATING = "SET_GENERATING";
export const RESET_KEY = "RESET_KEY";

// Actions
export function newWalletKeys(passphrase) {
  return {
    type: NEW_WALLET_KEYS,
    passphrase
  };
}

export function newWallet(account) {
  return {
    type: NEW_WALLET,
    wif: account.wif,
    address: account.address,
    passphrase: account.passphrase,
    encryptedWif: account.encryptedWif
  };
}

export function generating(bool) {
  return {
    type: SET_GENERATING,
    state: bool
  };
}

export function resetKey() {
  return {
    type: RESET_KEY
  };
}

// Reducer used for state necessary to generating a wallet
export default (
  state = {
    wif: null,
    address: null,
    passphrase: null,
    encryptedWif: null,
    generating: false
  },
  action
) => {
  switch (action.type) {
    case NEW_WALLET_KEYS:
      const newPrivateKey = generatePrivateKey();
      const newWif = getWIFFromPrivateKey(newPrivateKey);
      const encryptedWif = encrypt_wif(newWif, action.passphrase);
      const loadAccount = getAccountsFromWIFKey(newWif);
      return {
        ...state,
        wif: newWif,
        address: loadAccount[0].address,
        passphrase: action.passphrase,
        encryptedWif
      };
    case NEW_WALLET:
      return {
        ...state,
        wif: action.wif,
        address: action.address,
        passphrase: action.passphrase,
        encryptedWif: action.encryptedWif,
        generating: false
      };
    case SET_GENERATING:
      return { ...state, generating: action.state };
    case RESET_KEY:
      return {
        wif: null,
        address: null,
        passphrase: null,
        encryptedWif: null,
        generating: false
      };
    default:
      return state;
  }
};
