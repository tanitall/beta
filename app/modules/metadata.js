// Constants
export const SET_HEIGHT = "SET_HEIGHT";
export const SET_NETWORK = "SET_NETWORK";
export const SET_EXPLORER = "SET_EXPLORER";

// Actions
export function setNetwork(net) {
  const network = net === "MainNet" ? "MainNet" : "TestNet";
  return {
    type: SET_NETWORK,
    net: network
  };
}

export function setBlockHeight(blockHeight) {
  return {
    type: SET_HEIGHT,
    blockHeight
  };
}

export function setBlockExplorer(blockExplorer) {
  return {
    type: SET_EXPLORER,
    blockExplorer
  };
}

// reducer for metadata associated with Neon
export default (
  state = { blockHeight: 0, network: "TestNet", blockExplorer: "Neotracker" },
  action
) => {
  switch (action.type) {
    case SET_HEIGHT:
      return { ...state, blockHeight: action.blockHeight };
    case SET_EXPLORER:
      console.log(action.blockExplorer);
      return { ...state, blockExplorer: action.blockExplorer };
    case SET_NETWORK:
      return { ...state, network: action.net };
    default:
      return state;
  }
};
