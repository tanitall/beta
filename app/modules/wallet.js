// Constants
export const SET_BALANCE = "SET_BALANCE";
export const SET_MARKET_PRICE = "SET_MARKET_PRICE";
export const RESET_PRICE = "RESET_PRICE";
export const SET_TRANSACTION_HISTORY = "SET_TRANSACTION_HISTORY";

// Actions
export function setBalance(neo, gas, price, combined, gasPrice) {
  return {
    type: SET_BALANCE,
    Neo: neo,
    Gas: gas,
    price: price,
    combined: combined,
    gasPrice: gasPrice
  };
}

export function setMarketPrice(price) {
  return {
    type: SET_MARKET_PRICE,
    price: price
  };
}

export function resetPrice() {
  return {
    type: RESET_PRICE
  };
}

export function setTransactionHistory(transactions) {
  return {
    type: SET_TRANSACTION_HISTORY,
    transactions
  };
}

// reducer for wallet account balance
export default (
  state = {
    Neo: 0,
    Gas: 0,
    transactions: [],
    price: "--",
    combined: "--",
    gasPrice: "--"
  },
  action
) => {
  switch (action.type) {
    case SET_BALANCE:
      return {
        ...state,
        Neo: action.Neo,
        Gas: action.Gas,
        price: action.price,
        combined: action.combined,
        gasPrice: action.gasPrice
      };
    case RESET_PRICE:
      return { ...state, price: "--" };
    case SET_MARKET_PRICE: //current market price action type
      let currentPrice;
      if (action.price !== undefined) {
        currentPrice = action.price;
      } else {
        currentPrice = "--";
      }
      return { ...state, price: currentPrice };
    case SET_TRANSACTION_HISTORY:
      return { ...state, transactions: action.transactions };
    default:
      return state;
  }
};
