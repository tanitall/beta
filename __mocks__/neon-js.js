import { SEND_TRANSACTION } from '../app/modules/transactions';

const neonjs = jest.genMockFromModule('neon-js');

const promiseMockGen = (result, error = false) => {
  return jest.fn(() => {
    return new Promise((resolve, reject) => {
      if (error) reject();
      resolve(result);
    });
  });
};

const privateKey = 'L4AJ14CNaBWPemRJKC34wyZwbmxg33GETs4Y1F8uK7rRmZ2UHrJn';
const address = 'AM22coFfbe9N6omgL9ucFBLkeaMNg9TEyL';
const encryptedKey = '6PYUGtvXiT5TBetgWf77QyAFidQj61V8FJeFBFtYttmsSxcbmP4vCFRCWu';

neonjs.getTransactionHistory = promiseMockGen([]);
neonjs.getClaimAmounts = promiseMockGen({ available: 0, unavailable: 0 });
neonjs.getWalletDBHeight = promiseMockGen(586435);
neonjs.getBalance = promiseMockGen({ Neo: 1, Gas: 1 });
neonjs.getAPIEndpoint = jest.fn(() => 'http://testnet-api.wallet.cityofzion.io');
neonjs.decrypt_wif = jest.fn((wif) => {
  return new Promise((resolve, reject) => {
    if (!wif) reject();
    resolve(privateKey);
  });
});
neonjs.getAccountsFromWIFKey = jest.fn(() => [{ address }]);
neonjs.generatePrivateKey = jest.fn();
neonjs.getWIFFromPrivateKey = jest.fn(() => privateKey);
neonjs.encrypt_wif = jest.fn(() => encryptedKey);

module.exports = neonjs;
