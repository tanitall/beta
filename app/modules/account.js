import { getAccountsFromWIFKey } from "neon-js";

// Constants
export const LOGIN = "LOGIN";
export const LOGOUT = "LOGOUT";
export const SET_DECRYPTING = "SET_DECRYPTING";
export const SET_KEYS = "SET_KEYS";

// Actions
export function login(wif) {
  return {
    type: LOGIN,
    wif: wif
  };
}

export function logout() {
  return {
    type: LOGOUT
  };
}

export function decrypting(bool) {
  return {
    type: SET_DECRYPTING,
    state: bool
  };
}

export function setKeys(keys) {
  return {
    type: SET_KEYS,
    keys
  };
}

// Reducer that manages account state (account now = private key)
export default (
  state = {
    wif: null,
    address: null,
    loggedIn: false,
    redirectUrl: null,
    decrypting: false,
    accountKeys: []
  },
  action
) => {
  switch (action.type) {
    case LOGIN:
      let loadAccount;
      try {
        loadAccount = getAccountsFromWIFKey(action.wif)[0];
      } catch (e) {
        loadAccount = -1;
      }
      if (
        loadAccount === -1 ||
        loadAccount === -2 ||
        loadAccount === undefined
      ) {
        return { ...state, wif: action.wif, loggedIn: false };
      }
      return {
        ...state,
        wif: action.wif,
        address: loadAccount.address,
        loggedIn: true,
        decrypting: false
      };
    case LOGOUT:
      return {
        ...state,
        wif: null,
        address: null,
        loggedIn: false,
        decrypting: false
      };
    case SET_DECRYPTING:
      return { ...state, decrypting: action.state };
    case SET_KEYS:
      return { ...state, accountKeys: action.keys };
    default:
      return state;
  }
};
