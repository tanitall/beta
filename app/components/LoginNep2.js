import React, { Component } from "react";
import { connect } from "react-redux";
import { Link, browserHistory } from "react-router";
import { login, decrypting } from "../modules/account";
import CreateWallet from "./CreateWallet.js";
import { decrypt_wif } from "neon-js";
import Logo from "./Brand/LogoBlank";
// TODO: these event messages should be refactored from transactions
import { sendEvent, clearTransactionEvent } from "../modules/transactions";

const logo = require("../images/neon-logo2.png");

let wif_input;
let passphrase_input;

const onWifChange = (dispatch, history) => {
  if (passphrase_input.value.length < 8) {
    dispatch(sendEvent(false, "Wrong password"));
    setTimeout(() => dispatch(clearTransactionEvent()), 3000);
    return;
  }

  // TODO: changed back to only WIF login for now, getting weird errors with private key hex login
  dispatch(sendEvent(true, "Decrypting encoded key..."));
  setTimeout(() => {
    const encWifValue = wif_input.value;
    decrypt_wif(encWifValue, passphrase_input.value)
      .then(wif => {
        dispatch(login(wif));
        history.push("/dashboard");
        dispatch(clearTransactionEvent());
      })
      .catch(() => {
        dispatch(sendEvent(false, "Wrong password or invalid encrypted key"));
        setTimeout(() => dispatch(clearTransactionEvent()), 3000);
      });
  }, 500);
};

class LoginNep2 extends Component {
  render = () => {
    const dispatch = this.props.dispatch;
    return (
      <div>
        <div className="login-address-bk top-50">
          <div className="logo-top">
            <div className="row">
              <div className="center logobounce">
                <Logo width={140} />
              </div>
              <br />
              <h1 className="center">Welcome to Morpheus</h1>
            </div>

            <div className="row">
              <div className="col-xs-10 col-xs-offset-1">
                <div className="form-group">
                  <input
                    type="password"
                    className="trans-form"
                    placeholder="Enter your password"
                    ref={node => (passphrase_input = node)}
                  />
                </div>

                <div className="form-group">
                  <input
                    type="password"
                    className="trans-form"
                    placeholder="Enter your NEO encrypted key"
                    ref={node => (wif_input = node)}
                  />
                </div>
              </div>

              <div className="col-xs-10 col-xs-offset-1">
                <div
                  className="login-button"
                  onClick={e => onWifChange(dispatch, this.props.history)}
                >
                  Login
                </div>
              </div>

              <p className="center top-10 col-xs-10 col-xs-offset-1">
                Your encrypted private key and password are never shared and are
                only used to load your balance and transaction history from the
                blockchain.
              </p>
            </div>
            {this.props.decrypting === true ? (
              <div className="decrypting">Decrypting keys...</div>
            ) : (
              <div />
            )}
          </div>
        </div>
        <div className="row">
          <div className="dash-bar">
            <Link to="/create">
              <div className="dash-icon-bar">
                <div className="icon-border">
                  <div className="neo-icon" />
                </div>
                Create a NEO Address
              </div>
            </Link>
            <Link to="/">
              <div className="dash-icon-bar">
                <div className="icon-border">
                  <span className="glyphicon glyphicon-user" />
                </div>
                Open a Saved Wallet
              </div>
            </Link>
            <Link to="/LoginPrivateKey">
              <div className="dash-icon-bar">
                <div className="icon-border">
                  <span className="glyphicon glyphicon-qrcode" />
                </div>
                Login Via Private Key
              </div>
            </Link>
            <div
              className="dash-icon-bar"
              onClick={() => loadKeyRecovery(this.props.dispatch)}
            >
              <div className="icon-border">
                <span className="glyphicon glyphicon-paperclip" />
              </div>
              Login Via Recovery File
            </div>
          </div>
        </div>
      </div>
    );
  };
}

const mapStateToProps = state => ({
  decrypting: state.account.decrypting
});

LoginNep2 = connect(mapStateToProps)(LoginNep2);

export default LoginNep2;
