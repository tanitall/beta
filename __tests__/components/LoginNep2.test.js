import React from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import { createMemoryHistory } from 'history'
import LoginNep2 from '../../app/components/LoginNep2';
import { decrypt_wif } from 'neon-js';
import { SEND_TRANSACTION, CLEAR_TRANSACTION } from '../../app/modules/transactions';
import { LOGIN } from '../../app/modules/account';

jest.mock('neon-js');

const initialState = {
  account: {
    decrypting: false
  }
};

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore()(state);

  let wrapper;
  if (shallowRender) {
    wrapper = shallow(<LoginNep2 store={store} />);
  } else {
    const history = createMemoryHistory('/dashboard')
    wrapper = mount(
      <Provider store={store}>
        <LoginNep2 history={history} />
      </Provider>
    );
  }

  return {
    store,
    wrapper
  };
};

describe('LoginNep2', () => {
  test('renders without crashing', (done) => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
    done();
  });
  test('renders correctly with initial state', (done) => {
    const { wrapper } = setup(initialState, false);

    const passwordField = wrapper.find('input[type="password"]');
    const keyField = wrapper.find('input[type="text"]');
    const footerField = wrapper.find('#footer');

    expect(footerField.text()).toEqual('Created by Ethan Fast and COZ. Donations: Adr3XjZ5QDzVJrWvzmsTTchpLRRGSzgS5A');
    expect(passwordField.text()).toEqual('');
    expect(passwordField.html().includes('Enter your passphrase here')).toEqual(true);
    expect(keyField.text()).toEqual('');
    expect(keyField.html().includes('Enter your encrypted key here')).toEqual(true);
    done();
  });
  test('renders correctly when decrypting is true', (done) => {
    const newState = Object.assign({}, initialState, {
      account: {
        ...initialState.account,
        decrypting: true
      }
    });
    const { wrapper } = setup(newState, false);
    const decryptingText = wrapper.find('.decrypting');

    expect(decryptingText.text()).toEqual('Decrypting keys...');
    done();
  });
  test('the login button is working correctly with no passphrase', (done) => {
    const { wrapper, store } = setup();
    wrapper.dive().find('.loginButton').simulate('click');
    setTimeout(() => {
      const actions = store.getActions();
      expect(actions.length).toEqual(1);
      expect(actions[0]).toEqual({
        type: SEND_TRANSACTION,
        success: false,
        message: 'Passphrase too short'
      });
      done();
    }, 0)
  });
  test('the login button is working correctly with only a passphrase', (done) => {
    const { wrapper, store } = setup(initialState, false);
    const passwordField = wrapper.find('input[type="password"]');
    passwordField.instance().value = 'Th!s1$@FakePassphrase';
    passwordField.simulate('change')

    wrapper.find('.loginButton').simulate('click');
    setTimeout(() => {
      const actions = store.getActions();
      expect(actions.length).toEqual(2);
      expect(actions[0]).toEqual({
        type: SEND_TRANSACTION,
        success: true,
        message: 'Decrypting encoded key...'
      });
      expect(actions[1]).toEqual({
        type: SEND_TRANSACTION,
        success: false,
        message: 'Wrong passphrase or invalid encrypted key'
      });
      expect(decrypt_wif.mock.calls.length).toBe(1);
      expect(decrypt_wif.mock.calls[0][0]).toBe('');
      done();
    }, 510)
  });
  test('the login button is working correctly with key and passphrase', (done) => {
    const { wrapper, store } = setup(initialState, false);

    const passwordField = wrapper.find('input[type="password"]');
    passwordField.instance().value = 'Th!s1$@FakePassphrase';
    passwordField.simulate('change')

    const keyField = wrapper.find('input[type="text"]');
    keyField.instance().value = '6PYUGtvXiT5TBetgWf77QyAFidQj61V8FJeFBFtYttmsSxcbmP4vCFRCWu';
    keyField.simulate('change')

    wrapper.find('.loginButton').simulate('click');
    setTimeout(() => {
      const actions = store.getActions();
      expect(actions.length).toEqual(3);
      expect(decrypt_wif.mock.calls.length).toBe(2);
      expect(decrypt_wif.mock.calls[1][0]).toBe('6PYUGtvXiT5TBetgWf77QyAFidQj61V8FJeFBFtYttmsSxcbmP4vCFRCWu');
      expect(actions[0]).toEqual({
        type: SEND_TRANSACTION,
        success: true,
        message: 'Decrypting encoded key...'
      });
      expect(actions[1]).toEqual({
        type: LOGIN,
        wif: 'L4AJ14CNaBWPemRJKC34wyZwbmxg33GETs4Y1F8uK7rRmZ2UHrJn'
      });
      expect(actions[2]).toEqual({
        type: CLEAR_TRANSACTION
      });
      done();
    }, 510)
  });
});
