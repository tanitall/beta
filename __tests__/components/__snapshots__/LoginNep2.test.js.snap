// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginNep2 renders without crashing 1`] = `
<LoginNep2
  decrypting={false}
  dispatch={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
